File: app/src/main/java/com/hover/stax/database/DatabaseRepo.java
package com.hover.stax.database;

import android.annotation.SuppressLint;
import android.app.Application;
import android.content.Context;
import android.content.Intent;
import android.util.Log;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.hover.sdk.actions.HoverAction;
import com.hover.sdk.actions.HoverActionDao;
import com.hover.sdk.database.HoverRoomDatabase;
import com.hover.sdk.sims.SimInfo;
import com.hover.sdk.sims.SimInfoDao;
import com.hover.sdk.transactions.TransactionContract;
import com.hover.stax.R;
import com.hover.stax.channels.Channel;
import com.hover.stax.channels.ChannelDao;
import com.hover.stax.contacts.ContactDao;
import com.hover.stax.contacts.StaxContact;
import com.hover.stax.requests.Request;
import com.hover.stax.requests.RequestDao;
import com.hover.stax.requests.Shortlink;
import com.hover.stax.schedules.Schedule;
import com.hover.stax.schedules.ScheduleDao;
import com.hover.stax.transactions.StaxTransaction;
import com.hover.stax.transactions.TransactionDao;
import com.hover.stax.utils.Utils;
import com.hover.stax.utils.paymentLinkCryptography.Encryption;

import java.security.NoSuchAlgorithmException;
import java.util.List;

public class DatabaseRepo {
    private static String TAG = "DatabaseRepo";
    private ChannelDao channelDao;
    private HoverActionDao actionDao;
    private RequestDao requestDao;
    private ScheduleDao scheduleDao;
    private SimInfoDao simDao;
    private TransactionDao transactionDao;
    private ContactDao contactDao;

    private LiveData<List<Channel>> allChannels;
    private LiveData<List<Channel>> selectedChannels;

    private MutableLiveData<Request> decryptedRequest = new MutableLiveData<>();

    public DatabaseRepo(Application application) {
        AppDatabase db = AppDatabase.getInstance(application);
        channelDao = db.channelDao();
        transactionDao = db.transactionDao();
        contactDao = db.contactDao();
        requestDao = db.requestDao();
        scheduleDao = db.scheduleDao();

        HoverRoomDatabase sdkDb = HoverRoomDatabase.getInstance(application);
        actionDao = sdkDb.actionDao();
        simDao = sdkDb.simDao();

        allChannels = channelDao.getAllInAlphaOrder();
        selectedChannels = channelDao.getSelected(true);
    }

    // Channels
    public Channel getChannel(int id) {
        return channelDao.getChannel(id);
    }

    public LiveData<Channel> getLiveChannel(int id) {
        return channelDao.getLiveChannel(id);
    }

    public LiveData<List<Channel>> getAllChannels() {
        return allChannels;
    }

    public LiveData<List<Channel>> getChannels(int[] ids) {
        return channelDao.getChannels(ids);
    }

    public LiveData<List<Channel>> getChannelsByCountry(int[] channelIds, String countryCode) {
        return channelDao.getChannels(countryCode, channelIds);
    }

    public LiveData<List<Channel>> getSelected() {
        return selectedChannels;
    }

    public void update(Channel channel) {
        AppDatabase.databaseWriteExecutor.execute(() -> channelDao.update(channel));
    }

    // SIMs
    public List<SimInfo> getPresentSims() {
        return simDao.getPresent();
    }

    public List<SimInfo> getSims(String[] hnis) {
        return simDao.getPresentByHnis(hnis);
    }

    // Actions
    public HoverAction getAction(String public_id) {
        return actionDao.getAction(public_id);
    }

    public LiveData<HoverAction> getLiveAction(String public_id) {
        return actionDao.getLiveAction(public_id);
    }

    public LiveData<List<HoverAction>> getLiveActions(int[] channelIds, String type) {
        return actionDao.getLiveActions(channelIds, type);
    }

    public List<HoverAction> getTransferActions(int channelId) {
        return actionDao.getTransferActions(channelId);
    }

    public List<HoverAction> getActions(int channelId, String type) {
        return actionDao.getActions(channelId, type);
    }

    public List<HoverAction> getActions(int[] channelIds, String type) {
        return actionDao.getActions(channelIds, type);
    }

    public List<HoverAction> getActions(int[] channelIds, int recipientInstitutionId) {
        return actionDao.getActions(channelIds, recipientInstitutionId, HoverAction.P2P);
    }

    public LiveData<List<HoverAction>> getBountyActions() {
        return actionDao.getBountyActions();
    }

    // Transactions
    public LiveData<List<StaxTransaction>> getCompleteAndPendingTransferTransactions() {
        return transactionDao.getCompleteAndPendingTransfers();
    }
    public LiveData<List<StaxTransaction>> getTransactionsForAppReview() {
        return transactionDao.getTransactionsForAppReview();
    }

    public LiveData<List<StaxTransaction>> getBountyTransactions() {
        return transactionDao.getBountyTransactions();
    }

    public LiveData<List<StaxTransaction>> getCompleteTransferTransactions(int channelId) {
        return transactionDao.getCompleteAndPendingTransfers(channelId);
    }

    @SuppressLint("DefaultLocale")
    public LiveData<Double> getSpentAmount(int channelId, int month, int year) {
        return transactionDao.getTotalAmount(channelId, String.format("%02d", month), String.valueOf(year));
    }

    @SuppressLint("DefaultLocale")
    public LiveData<Double> getFees(int channelId, int year) {
        return transactionDao.getTotalFees(channelId, String.valueOf(year));
    }

    public StaxTransaction getTransaction(String uuid) {
        return transactionDao.getTransaction(uuid);
    }

    public void insertOrUpdateTransaction(Intent intent, Context c) {
        AppDatabase.databaseWriteExecutor.execute(() -> {
            try {
                StaxTransaction t = getTransaction(intent.getStringExtra(TransactionContract.COLUMN_UUID));
                StaxContact contact = intent.hasExtra(StaxContact.LOOKUP_KEY) ? getContact(intent.getStringExtra(StaxContact.LOOKUP_KEY)) : null;
                HoverAction a = intent.hasExtra(HoverAction.ID_KEY) ? getAction(intent.getStringExtra(HoverAction.ID_KEY)) : null;

                if (t == null) {
                    Utils.logAnalyticsEvent(c.getString(R.string.initializing_USSD_event_confirmed), c);
                    t = new StaxTransaction(intent, a, contact, c);
                    transactionDao.insert(t);
                }
                t.update(intent, a, contact, c);
                transactionDao.update(t);

                updateRequests(t, intent);
            } catch (Exception e) {
                Log.e(TAG, "error", e);
            }
        });
    }

    private void updateRequests(StaxTransaction t, Intent intent) {
        if (t.transaction_type.equals(HoverAction.RECEIVE)) {
            List<Request> rs = getRequests();
            for (Request r : rs) {
                StaxContact r_contact = getContact(r.requestee_ids);
                if (r_contact != null && r_contact.equals(new StaxContact(intent.getStringExtra("senderPhone")))) {
                    r.matched_transaction_uuid = t.uuid;
                    update(r);
                }
            }
        }
    }

    // Contacts
    public LiveData<List<StaxContact>> getAllContacts() {
        return contactDao.getAll();
    }

    public List<StaxContact> getContacts(String[] ids) {
        return contactDao.get(ids);
    }

    public LiveData<List<StaxContact>> getLiveContacts(String[] ids) {
        return contactDao.getLive(ids);
    }

    public StaxContact lookupContact(String lookupKey) {
        return contactDao.lookup(lookupKey);
    }

    public StaxContact getContact(String lookupKey) {
        return contactDao.lookup(lookupKey);
    }

    public LiveData<StaxContact> getLiveContact(String id) {
        return contactDao.getLive(id);
    }

    public StaxContact getContactFromPhone(String phone) {
        return contactDao.getContact(phone);
    }

    public void insertOrUpdate(StaxContact contact) {
        AppDatabase.databaseWriteExecutor.execute(() -> {
            if (getContact(contact.id) == null) {
                try {
                    contactDao.insert(contact);
                } catch (Exception e) {
                    Utils.logErrorAndReportToFirebase(TAG, "failed to insert contact", e);
                }
            } else
                contactDao.update(contact);
        });
    }

    public void update(StaxContact contact) {
        AppDatabase.databaseWriteExecutor.execute(() -> contactDao.update(contact));
    }

    // Schedules
    public LiveData<List<Schedule>> getFutureTransactions() {
        return scheduleDao.getLiveFuture();
    }

    public LiveData<List<Schedule>> getFutureTransactions(int channelId) {
        return scheduleDao.getLiveFutureByChannelId(channelId);
    }

    public Schedule getSchedule(int id) {
        return scheduleDao.get(id);
    }

    public void insert(Schedule schedule) {
        AppDatabase.databaseWriteExecutor.execute(() -> scheduleDao.insert(schedule));
    }

    public void update(Schedule schedule) {
        AppDatabase.databaseWriteExecutor.execute(() -> scheduleDao.update(schedule));
    }

    public void delete(Schedule schedule) {
        AppDatabase.databaseWriteExecutor.execute(() -> scheduleDao.delete(schedule));
    }

    // Requests
    public LiveData<List<Request>> getLiveRequests() {
        return requestDao.getLiveUnmatched();
    }
    public LiveData<List<Request>> getLiveRequests(int channelId) {
        return requestDao.getLiveUnmatchedByChannel(channelId);
    }

    public List<Request> getRequests() {
        return requestDao.getUnmatched();
    }

    public Request getRequest(int id) {
        return requestDao.get(id);
    }

    public LiveData<Request> decrypt(String encrypted, Context c) {
        if (decryptedRequest == null) { decryptedRequest = new MutableLiveData<>(); }
        decryptedRequest.setValue(null);
        String removedBaseUrlString = encrypted.replace(c.getString(R.string.payment_root_url, ""), "");

        //Only old stax versions contains ( in the link
        if (removedBaseUrlString.contains("(")) decryptRequestForOldVersions(removedBaseUrlString);
        else decryptRequest(removedBaseUrlString, c);
        return decryptedRequest;
    }

    private void decryptRequest(String param, Context c) {
        decryptedRequest.postValue(new Request(Request.decryptBijective(param, c)));
    }
    private void decryptRequestForOldVersions(String params) {
        try {
            Encryption e = Request.getEncryptionSettings().build();
            if (Request.isShortLink(params)) {
                params = new Shortlink(params).expand();
            }

            e.decryptAsync(params.replaceAll("[(]", "+"), new Encryption.Callback() {
                @Override
                public void onSuccess(String result) {
                    decryptedRequest.postValue(new Request(result));
                }

                @Override
                public void onError(Exception exception) {
                    Utils.logErrorAndReportToFirebase(TAG, "failed link decryption", exception);
                }
            });

        } catch (NoSuchAlgorithmException e) {
            Utils.logErrorAndReportToFirebase(TAG, "decryption failure", e);
        }
    }
    public void insert(Request request) {
        AppDatabase.databaseWriteExecutor.execute(() -> requestDao.insert(request));
    }

    public void update(Request request) {
        AppDatabase.databaseWriteExecutor.execute(() -> requestDao.update(request));
    }

    public void delete(Request request) {
        AppDatabase.databaseWriteExecutor.execute(() -> requestDao.delete(request));
    }

}

File: app/src/main/java/com/hover/stax/database/DatabaseRepo.kt
package com.hover.stax.database

import android.annotation.SuppressLint
import android.content.Context
import android.content.Intent
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import com.hover.sdk.actions.HoverAction
import com.hover.sdk.actions.HoverActionDao
import com.hover.sdk.database.HoverRoomDatabase
import com.hover.sdk.sims.SimInfo
import com.hover.sdk.sims.SimInfoDao
import com.hover.sdk.transactions.TransactionContract
import com.hover.stax.R
import com.hover.stax.account.Account
import com.hover.stax.account.AccountDao
import com.hover.stax.channels.Channel
import com.hover.stax.channels.ChannelDao
import com.hover.stax.contacts.ContactDao
import com.hover.stax.contacts.StaxContact
import com.hover.stax.requests.Request
import com.hover.stax.requests.RequestDao
import com.hover.stax.requests.Shortlink
import com.hover.stax.schedules.Schedule
import com.hover.stax.schedules.ScheduleDao
import com.hover.stax.transactions.StaxTransaction
import com.hover.stax.transactions.TransactionDao
import com.hover.stax.utils.DateUtils.lastMonth
import com.hover.stax.utils.Utils
import com.hover.stax.utils.paymentLinkCryptography.Encryption
import timber.log.Timber
import java.security.NoSuchAlgorithmException

class DatabaseRepo(db: AppDatabase, sdkDb: HoverRoomDatabase) {

    private val decryptedRequest: MutableLiveData<Request> = MutableLiveData()

    private val channelDao: ChannelDao = db.channelDao()
    private val actionDao: HoverActionDao = sdkDb.actionDao()
    private val requestDao: RequestDao = db.requestDao()
    private val scheduleDao: ScheduleDao = db.scheduleDao()
    private val simDao: SimInfoDao = sdkDb.simDao()
    private val transactionDao: TransactionDao = db.transactionDao()
    private val contactDao: ContactDao = db.contactDao()
    private val accountDao: AccountDao = db.accountDao()

    val allChannels: LiveData<List<Channel>> = channelDao.allInAlphaOrder
    val selected: LiveData<List<Channel>> = channelDao.getSelected(true)

    // Channels
    fun getChannel(id: Int): Channel {
        return channelDao.getChannel(id)
    }

    fun getLiveChannel(id: Int): LiveData<Channel> {
        return channelDao.getLiveChannel(id)
    }

    val channelsDataCount: Int
        get() = channelDao.dataCount

    fun getChannels(ids: IntArray?): LiveData<List<Channel>> {
        return channelDao.getChannels(ids)
    }

    fun getChannelsByCountry(channelIds: IntArray?, countryCode: String?): LiveData<List<Channel>> {
        return channelDao.getChannels(countryCode, channelIds)
    }

    fun getChannelsByCountry(countryCode: String?): List<Channel> {
        return channelDao.getChannels(countryCode)
    }

    fun update(channel: Channel?) {
        AppDatabase.databaseWriteExecutor.execute { channelDao.update(channel) }
    }

    // SIMs
    val presentSims: List<SimInfo>
        get() = simDao.present

    fun getSims(hnis: Array<String?>?): List<SimInfo> {
        return simDao.getPresentByHnis(hnis)
    }

    // Actions
    fun getAction(public_id: String?): HoverAction {
        return actionDao.getAction(public_id)
    }

    fun getLiveAction(public_id: String?): LiveData<HoverAction> {
        return actionDao.getLiveAction(public_id)
    }

    fun getLiveActions(channelIds: IntArray?, type: String?): LiveData<List<HoverAction>> {
        return actionDao.getLiveActions(channelIds, type)
    }

    fun getTransferActions(channelId: Int): List<HoverAction> {
        return actionDao.getTransferActions(channelId)
    }

    fun getActions(channelId: Int, type: String?): List<HoverAction> {
        return actionDao.getActions(channelId, type)
    }

    fun getActions(channelIds: IntArray?, type: String?): List<HoverAction> {
        return actionDao.getActions(channelIds, type)
    }

    fun getActions(channelIds: IntArray?, recipientInstitutionId: Int): List<HoverAction> {
        return actionDao.getActions(channelIds, recipientInstitutionId, HoverAction.P2P)
    }

    val bountyActions: LiveData<List<HoverAction>>
        get() = actionDao.bountyActions

    // Transactions
    val completeAndPendingTransferTransactions: LiveData<List<StaxTransaction>>?
        get() = transactionDao.getCompleteAndPendingTransfers()

    val bountyTransactions: LiveData<List<StaxTransaction>>?
        get() = transactionDao.bountyTransactions

    @SuppressLint("DefaultLocale")
    suspend fun hasTransactionLastMonth(): Boolean {
        return transactionDao.getTransactionCount(String.format("%02d", lastMonth().first), lastMonth().second.toString())!! > 0
    }

    fun getCompleteTransferTransactions(channelId: Int): LiveData<List<StaxTransaction>>? {
        return transactionDao.getCompleteAndPendingTransfers(channelId)
    }

    @SuppressLint("DefaultLocale")
    fun getSpentAmount(channelId: Int, month: Int, year: Int): LiveData<Double>? {
        return transactionDao.getTotalAmount(channelId, String.format("%02d", month), year.toString())
    }

    @SuppressLint("DefaultLocale")
    fun getFees(channelId: Int, year: Int): LiveData<Double>? {
        return transactionDao.getTotalFees(channelId, year.toString())
    }

    fun getTransaction(uuid: String?): StaxTransaction? {
        return transactionDao.getTransaction(uuid)
    }

    fun insertOrUpdateTransaction(intent: Intent, c: Context?) {
        AppDatabase.databaseWriteExecutor.execute {
            try {
                var t = getTransaction(intent.getStringExtra(TransactionContract.COLUMN_UUID))
                val a = getAction(intent.getStringExtra(HoverAction.ID_KEY))
                val channel = getChannel(a.channel_id)
                val contact = StaxContact.findOrInit(intent, channel.countryAlpha2, t, this)

                if (contact.accountNumber != null)
                    save(contact)

                if (t == null) {
                    c?.let { Utils.logAnalyticsEvent(c.getString(R.string.initializing_ussd_services), c) }
                    t = StaxTransaction(intent, a, contact, c)
                    transactionDao.insert(t)

                    t = transactionDao.getTransaction(t.uuid)
                }

                t!!.update(intent, a, contact, c)
                transactionDao.update(t)

                createAccounts(intent, t)
                updateRequests(t, contact)
            } catch (e: Exception) {
                Timber.e(e, "error")
            }
        }
    }

    private fun updateRequests(t: StaxTransaction?, contact: StaxContact) {
        if (t!!.transaction_type == HoverAction.RECEIVE) {
            val rs = requests
            for (r in rs) {
                if (r.requestee_ids.contains(contact.id) && Utils.getAmount(r.amount) == t.amount) {
                    r.matched_transaction_uuid = t.uuid
                    update(r)
                }
            }
        }
    }

    // Contacts
    val allContacts: LiveData<List<StaxContact>>
        get() = contactDao.all

    fun getContacts(ids: Array<String?>?): List<StaxContact> {
        return contactDao[ids]
    }

    fun getLiveContacts(ids: Array<String?>?): LiveData<List<StaxContact>> {
        return contactDao.getLive(ids)
    }

    fun lookupContact(lookupKey: String?): StaxContact {
        return contactDao.lookup(lookupKey)
    }

    fun getContact(id: String?): StaxContact? {
        return contactDao[id]
    }

    fun getContactByPhone(phone: String): StaxContact? {
        return contactDao.getByPhone("%$phone%")
    }

    fun getLiveContact(id: String?): LiveData<StaxContact> {
        return contactDao.getLive(id)
    }

    fun save(contact: StaxContact) {
        AppDatabase.databaseWriteExecutor.execute {
            if (getContact(contact.id) == null && contact.accountNumber != null) {
                try {
                    contactDao.insert(contact)
                } catch (e: Exception) {
                    Utils.logErrorAndReportToFirebase(TAG, "failed to insert contact", e)
                }
            } else contactDao.update(contact)
        }
    }

    // Schedules
    val futureTransactions: LiveData<List<Schedule>>
        get() = scheduleDao.liveFuture

    fun getFutureTransactions(channelId: Int): LiveData<List<Schedule>> {
        return scheduleDao.getLiveFutureByChannelId(channelId)
    }

    val transactionsForAppReview: LiveData<List<StaxTransaction>>?
        get() = transactionDao.transactionsForAppReview

    fun getSchedule(id: Int): Schedule {
        return scheduleDao[id]
    }

    fun insert(schedule: Schedule?) {
        AppDatabase.databaseWriteExecutor.execute { scheduleDao.insert(schedule) }
    }

    fun update(schedule: Schedule?) {
        AppDatabase.databaseWriteExecutor.execute { scheduleDao.update(schedule) }
    }

    fun delete(schedule: Schedule?) {
        AppDatabase.databaseWriteExecutor.execute { scheduleDao.delete(schedule) }
    }

    // Requests
    val liveRequests: LiveData<List<Request>>
        get() = requestDao.liveUnmatched

    fun getLiveRequests(channelId: Int): LiveData<List<Request>> {
        return requestDao.getLiveUnmatchedByChannel(channelId)
    }

    val requests: List<Request>
        get() = requestDao.unmatched

    fun getRequest(id: Int): Request {
        return requestDao[id]
    }

    fun decrypt(encrypted: String, c: Context): LiveData<Request> {
        decryptedRequest.value = null
        val removedBaseUrlString = encrypted.replace(c.getString(R.string.payment_root_url, ""), "")

        //Only old stax versions contains ( in the link
        if (removedBaseUrlString.contains("(")) decryptRequestForOldVersions(removedBaseUrlString)
        else decryptRequest(removedBaseUrlString, c)
        return decryptedRequest
    }

    private fun decryptRequest(param: String, c: Context) {
        decryptedRequest.postValue(Request(Request.decryptBijective(param, c)))
    }

    private fun decryptRequestForOldVersions(param: String) {
        var params = param
        try {
            val e = Request.getEncryptionSettings().build()
            if (Request.isShortLink(params)) {
                params = Shortlink(params).expand()
            }
            e.decryptAsync(params.replace("[(]".toRegex(), "+"), object : Encryption.Callback {
                override fun onSuccess(result: String) {
                    decryptedRequest.postValue(Request(result))
                }

                override fun onError(exception: Exception) {
                    Utils.logErrorAndReportToFirebase(TAG, "failed link decryption", exception)
                }
            })
        } catch (e: NoSuchAlgorithmException) {
            Utils.logErrorAndReportToFirebase(TAG, "decryption failure", e)
        }
    }

    fun insert(request: Request?) {
        AppDatabase.databaseWriteExecutor.execute { requestDao.insert(request) }
    }

    fun update(request: Request?) {
        AppDatabase.databaseWriteExecutor.execute { requestDao.update(request) }
    }

    fun delete(request: Request?) {
        AppDatabase.databaseWriteExecutor.execute { requestDao.delete(request) }
    }

    private fun createAccounts(intent: Intent, transaction: StaxTransaction) {
        Timber.e("Creating accounts")

        val data = intent.getSerializableExtra(TransactionContract.COLUMN_PARSED_VARIABLES) as HashMap<String, String>

        val accounts = mutableListOf<Account>()
        //TODO replace with action variable
        if (data.containsKey("userAccountList")) {
            //TODO parse out accounts
            Timber.e(data["userAccountList"])
        } else {
            val channel = getChannel(transaction.channel_id)

            with(channel) {
                val account = Account(name, name, logoUrl, accountNo, id, primaryColorHex, secondaryColorHex)
                accounts.add(account)
            }
        }
        Timber.e("Accounts - ${accounts.size}")

        saveAccounts(accounts)
    }

    val allAccounts: LiveData<List<Account>> = accountDao.getAllAccounts()

    suspend fun getAccounts(channelId: Int): List<Account> = accountDao.getAccounts(channelId)

    private fun getAccount(name: String, channelId: Int): Account? = accountDao.getAccount(name, channelId)

    private fun saveAccounts(accounts: List<Account>) {
        accounts.forEach { account ->
            val acct = getAccount(account.name, account.channelId)

            try {
                AppDatabase.databaseWriteExecutor.execute {
                    if (acct == null) {
                        Timber.e("Inserting account $account")
                        accountDao.insert(account)
                    } else {
                        Timber.e("Updating account $account")
                        accountDao.update(account)
                    }
                }
            } catch (e: Exception) {
                Utils.logErrorAndReportToFirebase(TAG, "failed to insert/update account", e)
            }
        }
    }

    fun insert(account: Account) {
        AppDatabase.databaseWriteExecutor.execute { accountDao.insert(account) }
    }

    fun update(account: Account) {
        AppDatabase.databaseWriteExecutor.execute { accountDao.update(account) }
    }

    companion object {
        private val TAG = DatabaseRepo::class.java.simpleName
    }
}